# Tutorial: Exploring coalescent variation across the genome
Mozes Blom, for questions [contact](mailto:mozes.blom@gmail.com)
___
## Requirements
* [Jupyter Lab](https://jupyterlab.readthedocs.io/en/stable/user/interface.html) and/or [Google Colab](https://colab.google/)
* Python 3+
    * [msprime](https://msprime.readthedocs.io/en/stable/)

___
## Introduction

The overarching aim of this practical is to visualise how genealogical history can vary across the genome and use computer simulations to explore how the rate of deep coalescence scales with past population dynamics. This tutorial can be done on its own but has been developed as a computer practical which demonstrates topics discussed in the lecture: **Reconciling gene tree - species tree discordance**, presented during the MSc course Phylogenetics in Ecology & Evolution at Potsdam University. This practical includes code snippets and examples from [Georgia Tsambos](https://github.com/gtsambos/2022-ts-workshops) and [Simon Martin](https://github.com/simonhmartin/tutorials/blob/master/topology_weighting/).

For this computer practical, we will use Python modules but a minimal understanding of Python code itself will be needed as all necessary code is written out in a Jupyter Notebook. However, for those unfamiliar with Jupyter Notebooks, please have a look at the general [README](https://github.com/MozesBlom/tutorials/tree/main/2025_PU_Phylo_Eco_Evol/README.md) before you get started.
